### PR TITLE
feat(nns): Voting rewards are now in proportion to total *potential* voting power.

### PR DESCRIPTION
**Pull request by daniel-wong-dfinity-org**
Whereas, before, we did not have a distinction between **potential** voting power vs. **deciding** voting power. When neurons refresh themselves, there is no difference between these. Whereas, when a neuron does not refresh itself recently enough, the amount of voting power in its ballots (its "deciding" voting power) is less than its potential voting power.

# Example

Suppose there are two neurons:

* A has 100 potential voting power, and refreshed recently, and as a result, its deciding voting power is also 100.
* B has 1000 potential voting power, but has not refreshed in a long time, and as a result, it's deciding voting power is just 500.

Then, the total potential voting power is 1100. If both neurons vote, then A gets 100 / 1100 of the reward purse, while B gets only 500 / 1100 of the reward purse, not 1000 / 1100.

You might also intuitively guess that the neurons get 100 / 600 and 500 / 600 of the reward purse, but this is (also) incorrect. The reason for this design is to keep the amount of rewards to refreshed neurons the same as before, not more. In other words, it should not be that some other neuron goes to sleep, and as a result, you get more rewards. It should only be that the sleeper neuron gets less.

# References

This is part of the "periodic confirmation" feature that was recently approved in proposal [132411][prop].

[prop]: https://dashboard.internetcomputer.org/proposal/132411

Closes https://dfinity.atlassian.net/browse/NNS1-3411.

[👈 Previous PR][prev] | [Next PR (tangent) 👉][next]

[prev]: https://github.com/dfinity/ic/pull/2375
[next]: https://github.com/dfinity/ic/pull/2470

### CODE REVIEW

File: rs/nns/governance/src/governance.rs
Comment (daniel-wong-dfinity-org) [inline rs/nns/governance/src/governance.rs:7055]: This was spun out to sum_weighted_voting_power in ./proposals/mod.rs. You may want to manually diff this deleted code with the body of that function. TBH, the changes there are not minimal.

**Review by daniel-wong-dfinity-org — COMMENTED** [review]
Hints for reviewers.

File: rs/nns/governance/src/governance.rs
Comment (max-dfinity) [inline rs/nns/governance/src/governance.rs:7026]: I'm not sure that moving this out of this particular function makes it clearer, or easier to understand.  

Additionally, the tests that were written were specific to using the new function, which means further refactoring would require writing new tests.  

Would it be possible to add a test that tests at a higher level, so that further changes are easier to make under test?

File: rs/nns/governance/src/proposals/mod.rs
Comment (aterga) [inline rs/nns/governance/src/proposals/mod.rs:68]: Reading this comment, I was a bit puzzled about why "highly unusual" refers to: 

Is the comment explaining why the code is unusual?

Or is the comment describing an unusual scenario? (I now this this is the case)

I suggest inverting the statement: Describe what is expected in a typical case, then say that sometimes this would not happen.

File: rs/nns/governance/src/proposals/mod.rs
Comment (aterga) [inline rs/nns/governance/src/proposals/mod.rs:70]: Why is it less important, and to whom?

File: rs/nns/governance/src/proposals/mod.rs
Comment (aterga) [inline rs/nns/governance/src/proposals/mod.rs:74]: IMHO, if you are adding comments about what these values mean, this implies it's time to make a struct that would carry that information properly. Maybe `WeightedVotingPower` would be a suitable name?

```suggestion
) -> (HashMap<NeuronId, WeightedVotingPower>) {
```

File: rs/nns/governance/src/proposals/mod.rs
Comment (aterga) [inline rs/nns/governance/src/proposals/mod.rs:126]: Is there a good enough reason why we're using floating point for reward calculation instead of `Decimal`?

File: rs/nns/governance/src/proposals/mod.rs
Comment (aterga) [inline rs/nns/governance/src/proposals/mod.rs:102]: If this is temporary code, should there be a TODO to eventually remove it?

File: rs/nns/governance/src/proposals/mod.rs
Comment (aterga) [inline rs/nns/governance/src/proposals/mod.rs:108]:
```suggestion
            // eligible neurons get an "empty" ballot when the proposal is first
```

File: rs/nns/governance/tests/proposals.rs
Comment (aterga) [inline rs/nns/governance/tests/proposals.rs:76]: Can we mark something with `Step 1: ...`?

File: rs/nns/governance/tests/proposals.rs
Comment (aterga) [inline rs/nns/governance/tests/proposals.rs:93]: How is the fact that it voted No reflected in this assertion?

**Review by aterga — APPROVED** [review]
Approving modulo comments

File: rs/nns/governance/src/governance.rs
Comment (daniel-wong-dfinity-org) [reply rs/nns/governance/src/governance.rs:7026]: Can you be more specific about what additional tests are needed? What is not already covered here?

File: rs/nns/governance/src/proposals/mod.rs
Comment (daniel-wong-dfinity-org) [reply rs/nns/governance/src/proposals/mod.rs]: I now realize you already approved -> I assume this means you leave it up to me. To me, this is not so complicated that it justifies a new struct. I think the mental weight of the indirection created by a new struct is too much for this.

File: rs/nns/governance/src/governance.rs
Comment (max-dfinity) [reply rs/nns/governance/src/governance.rs:7026]: A test at the level of distribute_rewards for the changes made, so that we can see the whole process is correct.

I was a bit surprised no changes to existing tests were needed, but I assume it's b/c there's a default behavior with existing data structures when certain values aren't filled in.

Comment (max-dfinity) [reply rs/nns/governance/src/governance.rs:7026]: Would it also be possible to implement the change in the existing function, and do the refactoring / moving in a separate PR?

File: rs/nns/governance/tests/proposals.rs
Comment (max-dfinity) [inline rs/nns/governance/tests/proposals.rs:199]: I don't see how this tests the new functionality... it just tests the proportion of rewards between the neurons, which would be the same whether or not the total_potential_voting_power was added, correct?

I feel like I may be missing something.

**Review by max-dfinity — APPROVED** [review]
Pre-approving, but if possible, please improve the test.

If that's not possible, I see it's covered in the next PR.

**Pull request merged**
